src/main/java/com/timgroup/statsd/NonBlockingDirectStatsDClient.java

```diff
@@ -0,0 +1,162 @@
+        private int offset = 0; // The index of the first value that has not been written
+
+        MultiValuedStatsDMessage(String aspect, Message.Type type, String[] tags, double sampleRate, long timestamp) {
+            super(aspect, type, tags);
+            this.sampleRate = sampleRate;
+            this.timestamp = timestamp;
+        }
+
+        @Override
+        public final boolean canAggregate() {
+            return false;
+        }
+
+        @Override
+        public final void aggregate(Message message) {
+        }
+
+        @Override
+        public final boolean writeTo(StringBuilder builder, int capacity, String containerID) {
+            int metadataSize = metadataSize(builder, containerID);
+            writeHeadMetadata(builder);
+            boolean partialWrite = writeValuesTo(builder, capacity - metadataSize);
+            writeTailMetadata(builder, containerID);
+            return partialWrite;
+
+        }
+
+        private int metadataSize(StringBuilder builder, String containerID) {
+            if (metadataSize == -1) {
+                final int previousLength = builder.length();
+                final int previousEncodedLength = builder.toString().getBytes(UTF_8).length;
+                writeHeadMetadata(builder);
+                writeTailMetadata(builder, containerID);
+                metadataSize = builder.toString().getBytes(UTF_8).length - previousEncodedLength;
+                builder.setLength(previousLength);
+            }
+            return metadataSize;
+        }
+
+        private void writeHeadMetadata(StringBuilder builder) {
+            builder.append(prefix).append(aspect);
+        }
+
+        private void writeTailMetadata(StringBuilder builder, String containerID) {
+            builder.append('|').append(type);
+            if (!Double.isNaN(sampleRate)) {
+                builder.append('|').append('@').append(format(SAMPLE_RATE_FORMATTER, sampleRate));
+            }
+            if (timestamp != 0) {
+                builder.append("|T").append(timestamp);
+            }
+            tagString(tags, builder);
+            if (containerID != null && !containerID.isEmpty()) {
+                builder.append("|c:").append(containerID);
+            }
+
+            builder.append('\n');
+        }
+
+        private boolean writeValuesTo(StringBuilder builder, int remainingCapacity) {
+            int maxLength = builder.length() + remainingCapacity;
+
+            // Add at least one value
+            builder.append(':');
+            writeValueTo(builder, offset);
+            int previousLength = builder.length();
+
+            // Add remaining values up to the max length
+            for (int i = offset + 1; i < lengthOfValues(); i++) {
+                builder.append(':');
+                writeValueTo(builder, i);
+                if (builder.length() > maxLength) {
+                    builder.setLength(previousLength);
+                    offset += i;
```

Comment:
I could be missing something, but this looks like it will skip values for chunks after second one: let's say we start with offset=3 and write 4 values at positions 3, 4, 5 and 6 before hitting overflow at i=7. This sets offset to 12 (3+7), loosing samples 7-11.

Reply:
You are right, I have updated the test and the implementation to fix that in [d4f3cf8](https://github.com/DataDog/java-dogstatsd-client/pull/235/commits/d4f3cf8a08d1e96f01d23fbb5434cf6882c2a59c)